test/e2e/ipam/ipamhelper.go

```diff
@@ -0,0 +1,373 @@
+	for _, variable := range append(additionalIPVariableNames, controlPlaneEndpointVariable) {
+		ip, err := ipamService.claimIPAddress(ctx)
+		Expect(err).ToNot(HaveOccurred())
+		Byf("Setting clusterctl variable %s to %s", variable, ip)
+		variables[variable] = ip
+	}
+
+	// Create a new clusterctl config file based on the passed file and add the new variables for the IPs.
+	modifiedClusterctlConfigPath := fmt.Sprintf("%s-%s.yaml", strings.TrimSuffix(clusterctlConfigPath, ".yaml"), rand.String(16))
+	Byf("Writing a new clusterctl config to %s", modifiedClusterctlConfigPath)
+	copyAndAmendClusterctlConfig(ctx, copyAndAmendClusterctlConfigInput{
+		ClusterctlConfigPath: clusterctlConfigPath,
+		OutputPath:           modifiedClusterctlConfigPath,
+		Variables:            variables,
+	})
+
+	return modifiedClusterctlConfigPath, ipamService.cleanup
+}
+
+// Teardown lists all IPAddressClaims matching the passed labels and deletes the IPAddressClaim
+// if there are no VirtualMachines in vCenter using the IP address.
+func Teardown(ctx context.Context, e2eIPAMKubeconfig, folderName string, vSphereClient *govmomi.Client) error {
+	if e2eIPAMKubeconfig == "" {
+		By("Skipping IPAM Teardown because no kubeconfig is set")
+		return nil
+	}
+
+	ipamClient, err := getClientSet(e2eIPAMKubeconfig)
+	if err != nil {
+		return err
+	}
+
+	virtualMachineIPAddresses, err := getVirtualMachineIPAddresses(ctx, folderName, vSphereClient)
+	if err != nil {
+		return err
+	}
+	// List all IPAddressClaims created matching the labels.
+	ipAddressClaims := &ipamv1.IPAddressClaimList{}
+	if err := ipamClient.List(ctx, ipAddressClaims,
+		client.MatchingLabels(getIPAddressClaimLabels()),
+		client.InNamespace(metav1.NamespaceDefault),
+	); err != nil {
+		return err
+	}
+
+	ipAddressClaimsToDelete := []ipamv1.IPAddressClaim{}
+	// Collect errors and skip these ip address claims, but report at the end.
+	var errList []error
+
+	ip := &ipamv1.IPAddress{}
+	for _, ipAddressClaim := range ipAddressClaims.Items {
+		if err := ipamClient.Get(ctx, client.ObjectKey{Namespace: ipAddressClaim.GetNamespace(), Name: ipAddressClaim.Status.AddressRef.Name}, ip); err != nil {
+			errList = append(errList, errors.Wrapf(err, "Getting IPAddress for IPAddressClaim %s/%s", ipAddressClaim.GetNamespace(), ipAddressClaim.GetName()))
+			continue
+		}
+
+		// Skip deletion if there is still a virtual machine which refers this IP address.
+		if virtualMachineIPAddresses[ip.Spec.Address] {
+			continue
+		}
+
+		ipAddressClaimsToDelete = append(ipAddressClaimsToDelete, ipAddressClaim)
+	}
+
+	for _, ipAddressClaim := range ipAddressClaimsToDelete {
+		ipAddressClaim := ipAddressClaim
+		Byf("Deleting IP Addressclaim %s/%s", ipAddressClaim.GetNamespace(), ipAddressClaim.GetName())
+		if err := ipamClient.Delete(ctx, &ipAddressClaim); err != nil {
+			errList = append(errList, errors.Wrapf(err, "Deleting IPAddressClaim %s/%s", ipAddressClaim.GetNamespace(), ipAddressClaim.GetName()))
```

Comment:
```suggestion
			errList = append(errList, errors.Wrapf(err, "Deleting IPAddressClaim %s", klog.KObj(ipAddressClaim))
```

Not super important here in test code, just wanted to bring it up. Usually this is the idiomatic way that we're doing this (also very close to what we use when we add it as a k/v pair).

The resulting log here will be the same though